pkg/storer/reserve.go

```diff
@@ -213,6 +213,15 @@ func (db *DB) evictionWorker(ctx context.Context) {
 			// it uses the same one as reserve eviction.
 			defer db.events.Trigger(reserveUnreserved)
 
+			// this trigger ensures that if the expiry triggers were swallowed
+			// after we checked for expired batches, we re-check the batches
+			// again. It should not be fired if there are no more batches to evict.
+			defer func() {
+				_ = time.AfterFunc(1*time.Second, func() {
```

Comment:
this is probably due to the early exit if the semaphore cannot be acquired here https://github.com/ethersphere/bee/pull/4237/files#diff-acdf952e80a8279dfab4acea0ff533558f2857121497d2a2a343d74058ddef7dR175

by changing the interactions with the semaphore, this after func won't be needed.

Reply:
since the internals of the cleanupExpired func is inside a goroutine anyways, waiting for the semaphore could work, eg calling `Acquire`.

Reply:
if the above is not ideal, you can also stick a `db.events.Trigger(batchExpiry)` here https://github.com/ethersphere/bee/pull/4237/files/53efc7eff387949ca0b290249807abb4f2b0b12d#diff-acdf952e80a8279dfab4acea0ff533558f2857121497d2a2a343d74058ddef7dR176

Reply:
if the above is not ideal, you can add a forever loop inside the goroutine of `cleanupExpired`
```go
for {
 batchesToEvict, err := db.getExpiredBatches()
 if len(batchesToEvict) == 0 {
  break
 }
}
```

Reply:
If the main worker waits to acquire the semaphore, then it cannot process unreserve/cleanup-related signals till the expiration finishes. So the idea was the main worker should be unblocked all the time. Secondly, if we start goroutines for each trigger waiting to acquire, then we might overwhelm the CPU. Also if we trigger before returning, it will cause excessive checking and creation of goroutines.

Reply:
what about the loop idea?

Reply:
Okay this is better

Reply:
> what about the loop idea?

The loop still has a race condition because the zero is detected before the semaphore is released, so if a trigger fires and is ignored between the loop end and the defer semaphore release, that trigger will still be ignored.

Reply:
> > what about the loop idea?
> 
> The loop still has a race condition because the zero is detected before the semaphore is released, so if a trigger fires and is ignored between the loop end and the defer semaphore release, that trigger will still be ignored.

So I feel this is better than the time.After idea. We would proactively remove instead of waiting for the signal to fire it again. If we miss out on entries due to the race, we always have either cleanup or new expirations that can trigger us again.